### PR TITLE
[REVIEW] MST core solver 

### LINKED ISSUE #52: [FEA] MST kernels
# Minimum Spanning Tree (MST)
 
## Input

*   Format: CSR - single GPU
*   Weighted: Yes
*   Directed: No
*   Size
    *   V=1M-2B
    *   E=1M-2B
    *   EF > 16
* Need to support multiple components

## Output

Set of edges present in the MST. A boolean array of size E where mst[i] is true if edge [i] is in the MST, false otherwise. 

## SW

Both cuML and cuGraph need this, RAFT would be a natural place to host the device code. 

*   **cuML** can leverage this as part of hierarchical clustering. KNN forms a reduced neighborhood graph that is used to construct the MST (as an optimization to shrink the space of total potential edges). In other words, the Sparse KNN would provide the graph that is then used to construct the MST instead of the KNN using the MST directly.
*   **cuGraph** can add python bindings and expose it as a graph algo in the API.

## Litterature

- [Fast Minimum Spanning Tree for Large Graphs on the GPU](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.536.874&rep=rep1&type=pdf)
- [Large Graph Algorithms for Massively Multithreaded Architectures](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.417.2999&rep=rep1&type=pdf) slower but simpler with all kernels documented.
- [kNN-Boruvka-GPU: A Fast & Scalable MST Construction from KNN Graphs on GPU](https://drive.google.com/drive/u/1/folders/1jzVInOgWquqi7xUWXAO00m2ImwmcJS8d) The benchmarks demonstrated to scale to larger numbers of vertices/edges. The fastest implementation in this paper is “Fast minimum spanning tree for large graphs on the GPU” but scalability problems.
- “Fast Euclidean Minimum Spanning Tree: Algorithm, Analysis, & Applications.
- "Design & Implementation of Cover Tree Algorithm on CUDA-Compatible GPU” 

### Other considerations

If we reduce this to knn graphs then we can further optimize but we lose a lot on the generic aspect. In this case we would have CSR of fixed degree and we know the edge with the shortest distance is going to be the first edge for each row of the CSR for instance.

---

Thanks to @cjnolet for helping to scope this new feature

### PR DESCRIPTION
**Pull request by afender**
co-authored with @divyegala 
closes #52 

Needed to unblock:
- https://github.com/rapidsai/cuml/issues/2727 and https://github.com/rapidsai/cuml/issues/3099
- https://github.com/rapidsai/cugraph/pull/1276 

----

### Parallel Borůvka's algorithm with support for:
- disconnected components (minimum spanning **forest** problem)
- negative edge weights (and **maximum** spanning tree/forest problem)
- equal weights (unlike NetworkX)
- multigraph
- singletons
- explicit zero weights are legit weights

Tested on 100M+ edges

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+// Finds the minimum edge from each vertex to the lowest color
+template <typename vertex_t, typename edge_t, typename weight_t>
+void MST_solver<vertex_t, edge_t, weight_t>::min_edge_per_vertex() {
+  thrust::fill(min_edge_color.begin(), min_edge_color.end(),
+               std::numeric_limits<weight_t>::max());
+
+  int n_threads = 32;
+
+  vertex_t* color_index_ptr = thrust::raw_pointer_cast(color_index.data());
```
Comment (ChuckHastings) [inline cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh:306]: An observation... I have been using the `.get()` method on the pointer to retrieve this value:

```vertex_t* color_index_ptr = color_index.data().get();```

feels more pleasant to me.  I believe they are equivalent constructs.

**Review by ChuckHastings — CHANGES_REQUESTED** [review]

File: cpp/include/raft/sparse/mst/detail/mst_kernels.cuh [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+#include <raft/device_atomics.cuh>
+
+namespace raft {
+namespace mst {
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+__global__ void kernel_min_edge_per_vertex(
+  const edge_t* offsets, const edge_t* indices, const weight_t* weights,
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/include/raft/sparse/mst/detail/mst_kernels.cuh]: yes they should :)  29605b5

File: cpp/include/raft/sparse/mst/mst_solver.cuh [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  raft::Graph_COO<vertex_t, edge_t, weight_t> solve();
+
+  ~MST_solver() {}
+
+ private:
+  raft::handle_t const& handle;
+  cudaStream_t stream;
+
+  //CSR
+  const edge_t* offsets;
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/include/raft/sparse/mst/mst_solver.cuh:56]: reconciled in 9fe2434

File: cpp/test/mst.cu [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+#include <raft/cudart_utils.h>
+#include <raft/handle.hpp>
+
+#include <raft/sparse/mst/mst.cuh>
+
+#include "test_utils.h"
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+struct CSRHost {
+  std::vector<vertex_t> offsets;
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/test/mst.cu]: absolutely! 29605b5

File: cpp/test/mst.cu [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+  return total_weight;
+}
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+class MSTTest
+  : public ::testing::TestWithParam<CSRHost<vertex_t, edge_t, weight_t>> {
+ protected:
+  raft::Graph_COO<vertex_t, edge_t, weight_t> mst_sequential() {
+    vertex_t *offsets = static_cast<vertex_t *>(csr_d.offsets.data());
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/test/mst.cu]: 9fe2434

File: cpp/test/mst.cu [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    edge_t *indices = static_cast<edge_t *>(csr_d.indices.data());
+    weight_t *weights = static_cast<weight_t *>(csr_d.weights.data());
+
+    v = static_cast<vertex_t>((csr_d.offsets.size() / sizeof(weight_t)) - 1);
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/test/mst.cu]: 29605b5

File: cpp/test/mst.cu [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    EXPECT_LE(result.n_edges, 2 * v - 2);
+
+    return result;
+  }
+
+  void SetUp() override {
+    csr_h =
+      ::testing::TestWithParam<CSRHost<vertex_t, edge_t, weight_t>>::GetParam();
+
+    csr_d.offsets = rmm::device_buffer(csr_h.offsets.data(),
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/test/mst.cu:152]: 29605b5

File: cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh (hunk 1 above)
Comment (afender) [reply cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh]: It is nicer indeed, thanks for the suggestion I removed all `raw_pointer_cast` in c25a559

**Review by ChuckHastings — APPROVED** [review]

**Review by BradReesWork — APPROVED** [review]

**Pull request merged**